velox/functions/prestosql/Split.cpp

```diff
@@ -0,0 +1,263 @@
+                                  .argumentType("varchar")
+                                  .argumentType("varchar")
+                                  .argumentType(limitType)
+                                  .build());
+    }
+    return signatures;
+  }
+
+ private:
+  /**
+   * The function calls the 'typed' template function based on the type of the
+   * 'limit' argument.
+   */
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::Expr* /*caller*/,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // If no 'limit' specified we just pick some int type.
+    // Should we have two passes, with limit and without?
+    TypeKind limitType =
+        (args.size() == 2) ? TypeKind::INTEGER : args[2]->typeKind();
+    switch (limitType) {
+      case TypeKind::INTEGER:
+        applyTyped<int32_t>(rows, args, context, result);
+        return;
+      case TypeKind::BIGINT:
+        applyTyped<int64_t>(rows, args, context, result);
+        return;
+      default:
+        VELOX_FAIL(
+            "Unsupported type for 'limit' argument of 'split' function: {}",
+            mapTypeKindToName(limitType));
+    }
+  }
+
+  template <typename I>
+  void applyTyped(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    // Get the decoded vectors out of arguments.
+    const bool noLimit = (args.size() == 2);
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* delims = decodedArgs.at(1);
+    DecodedVector* limits = noLimit ? nullptr : decodedArgs.at(2);
+
+    // Build the result vector (arrays of strings) using Array Builder.
+    ArrayBuilder<Varchar> builder(
+        rows.end(), 3 * rows.countSelected(), context->pool());
+
+    // Optimization for the (flat, const, const) case.
+    if (strings->isIdentityMapping() and delims->isConstantMapping() and
+        (noLimit or limits->isConstantMapping())) {
+      const auto* rawStrings = strings->data<StringView>();
+      const auto delim = delims->valueAt<StringView>(0);
+
+      if (noLimit) {
+        const I limit = std::numeric_limits<I>::max();
+        rows.applyToSelected([&](vector_size_t row) {
+          applyInner<false, I>(rawStrings[row], delim, limit, row, builder);
+        });
+      } else {
+        const I limit = limits->valueAt<I>(0);
+        rows.applyToSelected([&](vector_size_t row) {
+          applyInner<true, I>(rawStrings[row], delim, limit, row, builder);
+        });
+      }
+
+      // Ensure that our result elements vector uses the same string buffer as
+      // the input vector of strings.
+      builder.setStringBuffers(
+          strings->base()->as<FlatVector<StringView>>()->stringBuffers());
+    } else if (
+        strings->isConstantMapping() and delims->isConstantMapping() and
```

Comment:
Won't the all-constant case be optimized automatically by the logic that cals this function?